scRNA-seq-advanced/04-gene_set_enrichment_analysis.Rmd

```diff
@@ -0,0 +1,286 @@
+
+```{r hallmark_head, live = TRUE}
+head(hs_hallmarks_df)
+```
+
+We can see that the gene sets from `msigdbr` have Ensembl gene identifiers associated with them, so we don't need to do any conversion.
+However, we'll need to pass the correct column to the function that runs GSEA.
+
+One good thing about Ensembl gene identifiers is that they are less likely to be duplicated than, for example, gene symbols.
+(Multiple Ensembl gene identifiers can map to the same symbol.)
+
+The GSEA approach requires on discriminating between genes that are in a gene set and those that are not.
+Practically speaking, gene sets are just collections of gene identifiers!
+When the function we use for GSEA pre-ranked gets a list with duplicated gene identifiers, it can produce unexpected results.
+So, let's check for duplicates in the data frame of DESeq2 results.
+
+```{r check_duplicates, live = TRUE}
+any(duplicated(deseq_df$ensembl_id))
+```
+
+There are no duplicates for us to worry about!
+
+### Pre-ranked list
+
+The `GSEA()` function takes a pre-ranked (sorted) named vector of statistics, where the names in the vector are gene identifiers.
+This is step 1 -- gene-level statistics.
+
+```{r lfc_vector}
+lfc_vector <- deseq_df |>
+  # Extract a vector of `log2FoldChange` named by `ensembl_id`
+  dplyr::pull(log2FoldChange, name = ensembl_id)
+lfc_vector <- sort(lfc_vector, decreasing = TRUE)
+```
+
+Let's look at the top ranked values.
+
+```{r head_lfc, live = TRUE}
+# Look at first entries of the log fold change vector
+head(lfc_vector)
+```
+
+And the bottom of the list.
+
+```{r tail_lfc, live = TRUE}
+# Look at the last entries of the log fold change vector
+tail(lfc_vector)
+```
+
+## Run GSEA
+
+Now for the analysis!
+
+We can use the `GSEA()` function to perform GSEA with any generic set of gene sets, but there are several functions for using specific, commonly used gene sets (e.g., `gseKEGG()`).
+
+```{r run_gsea}
+gsea_results <- GSEA(geneList = lfc_vector,  # ordered ranked gene list
+                     minGSSize = 25,  # minimum gene set size
+                     maxGSSize = 500,  # maximum gene set set
+                     pvalueCutoff = 0.05,
+                     pAdjustMethod = "BH",  # correction for multiple hypothesis testing
+                     TERM2GENE = dplyr::select(hs_hallmarks_df,
+                                               gs_name,
+                                               ensembl_gene))  # pass the correct identifier column
+```
+Let's take a look at the GSEA results.
+
+```{r view_gsea, live = TRUE, eval = FALSE}
+View(gsea_results@result |>
+       dplyr::arrange(dplyr::desc(NES))
+)
+```
+
+Normalized enrichment scores (NES) are enrichment scores that are scaled to make gene sets that contain different number of genes comparable.
+
+Pathways with significant, highly positive NES are enriched in ERMS myoblasts, whereas pathways with significant, highly negative NES are enriched in ARMS myoblasts.
```

Comment:
Before figuring out who to request for a full review, I'm going to ask @jashapiro to take a look at this interpretation as the instructor of the `03` notebook in the upcoming workshop.

Reply:
This is correct. The scores are relative to ARMS, so:
- Positive values ---> ERMS is higher than ARMS ---> Enriched for ERMS.
- Negative values ---> ERMS is lower than ARMS ---> Enriched for ARMS.